presto-orc/src/main/java/com/facebook/presto/orc/reader/MapDirectSelectiveStreamReader.java

```diff
@@ -0,0 +1,645 @@
+        this.nullsAllowed = filter == null || filter.testNull();
+        this.nonNullsAllowed = filter == null || filter.testNonNull();
+
+        List<StreamDescriptor> nestedStreams = streamDescriptor.getNestedStreams();
+
+        Optional<Type> keyOutputType = outputType.map(MapType.class::cast).map(MapType::getKeyType);
+        Optional<Type> valueOutputType = outputType.map(MapType.class::cast).map(MapType::getValueType);
+
+        Map<Subfield, TupleDomainFilter> keyFilter = makeKeyFilter(nestedStreams.get(0).getStreamType(), requiredSubfields)
+                .map(f -> ImmutableMap.of(new Subfield("c"), f))
+                .orElse(ImmutableMap.of());
+
+        this.keyReader = SelectiveStreamReaders.createStreamReader(nestedStreams.get(0), keyFilter, keyOutputType, ImmutableList.of(), hiveStorageTimeZone, systemMemoryContext.newAggregatedMemoryContext());
+        this.valueReader = SelectiveStreamReaders.createStreamReader(nestedStreams.get(1), ImmutableMap.of(), valueOutputType, ImmutableList.of(), hiveStorageTimeZone, systemMemoryContext.newAggregatedMemoryContext());
+    }
+
+    private static Optional<TupleDomainFilter> makeKeyFilter(OrcType.OrcTypeKind orcType, List<Subfield> requiredSubfields)
+    {
+        if (requiredSubfields.isEmpty()) {
+            return Optional.empty();
+        }
+
+        switch (orcType) {
+            case BYTE:
+            case SHORT:
+            case INT:
+            case LONG: {
+                long[] requiredIndices = requiredSubfields.stream()
+                        .map(Subfield::getPath)
+                        .map(path -> path.get(0))
+                        .map(Subfield.LongSubscript.class::cast)
+                        .mapToLong(Subfield.LongSubscript::getIndex)
+                        .toArray();
+
+                if (requiredIndices.length == 0) {
+                    return Optional.empty();
+                }
+
+                if (requiredIndices.length == 1) {
+                    return Optional.of(BigintRange.of(requiredIndices[0], requiredIndices[0], false));
+                }
+
+                return Optional.of(BigintValues.of(requiredIndices, false));
+            }
+            case STRING:
+            case CHAR:
+            case VARCHAR: {
+                byte[][] requiredIndices = requiredSubfields.stream()
+                        .map(Subfield::getPath)
+                        .map(path -> path.get(0))
+                        .map(Subfield.StringSubscript.class::cast)
+                        .map(Subfield.StringSubscript::getIndex)
+                        .map(String::getBytes)
+                        .toArray(byte[][]::new);
+
+                if (requiredIndices.length == 0) {
+                    return Optional.empty();
+                }
+
+                if (requiredIndices.length == 1) {
+                    return Optional.of(BytesRange.of(requiredIndices[0], false, requiredIndices[0], false, false));
```

Comment:
for my understanding, why can't we just use BigIntValue even for a single value use case

Reply:
@bhhari `BigintValues` class doesn't support single value case (for performance reasons):

```
        private BytesValues(byte[][] values, boolean nullAllowed)
        {
            super(true, nullAllowed);

            requireNonNull(values, "values is null");
            checkArgument(values.length > 1, "values must contain at least 2 entries");
```